src/Mailer/Message.php

```diff
@@ -999,7 +999,7 @@ protected function formatAddress(array $address): array
                 $return[] = $email;
             } else {
                 $encoded = $this->encodeForHeader($alias);
-                if (preg_match('/[^a-z0-9+\-\\=? ]/i', $encoded)) {
+                if ($encoded === $alias && preg_match('/[^a-z0-9+\-\\=? ]/i', $encoded)) {
```

Comment:
Unfortunately this reintroduces the vulnerability. `mb_encode_mimeheader` sometimes splits the encoded string into encoded and not encoded pieces: "test ?test" results in "test =?UTF-8?B?P3Rlc3Q=?=". I don't know how the resulting string would be handled correctly according to the RFCs.

Looking at the _PHPMailer_ project it seems they have their own encoding method that always encodes the whole string, see https://github.com/PHPMailer/PHPMailer/blob/e88da8d679acc3824ff231fdc553565b802ac016/src/PHPMailer.php#L3591C21-L3591C39.

Reply:
Encoding the entire address alias could be a solution for us as well then. I re-introduced this comparison because we would otherwise be putting encoded values in quoted strings which would then be treated as literal values and not encoded content.

Reply:
I took a look at the PHPMailer implementation and it doesn't correctly handle a scenario that is important to the CakePHP community which is encoding email headers in ISO-2022-JP encoding. I also re-read the RFC 2047 and my initial interpretation was wrong. I've reverted my changes.